Monitoring guide with Prometheus (#144)
platformatic/platformatic: docs/guides/monitoring.md

# Monitoring with Prometheus and Grafana

[Prometheus](https://auth0.com/) is open source system and alerting toolkit for monitoring and alerting. It's a time series database that collects metrics from configured targets at given intervals, evaluates rule expressions, displays the results, and can trigger alerts if some condition is observed to be true.
[Grafana](https://grafana.com/oss/grafana/) is an open source visualization and analytics software. 

It's a pretty common solution to use Prometheus to collect and store monitoring data, and Grafana to visualize it.

Platformatic can be configured to expose Prometheus metrics: 

```json
...
  "metrics": {
    "port": 9091,
    "auth": {
      "username": "platformatic",
      "password": "mysecret"
    }
  }
...
```

In this case, we are exposing the metrics on port 9091 (defaults to `9090`), and we are using basic authentication to protect the endpoint.
We can also specify the IP address to bind to (defaults to `0.0.0.0`). 

> Note that the metrics endpoint is exposed on a different port than the APIs. This is because if you want to test the integration running both Prometheus and Platformatic on the same host, Prometheus starts on `9090` port too.  

Reviewer: ```suggestion
Note that the metrics endpoint is exposed on a different port than the Platformatic DB APIs. This is because if you want to test the integration running both Prometheus and Platformatic on the same host, Prometheus starts on `9090` port too.  
```

Author: This is actually wrong @simonplend . Platformatic DB APIis on `3042`, Platformatic METRICS endpoint is on `9090`, which is the same as `Prometheus`. So Cahnging this, but specifying that we are talking about the Platformatic metric endpoint.

> All the configurations are optional, to use the default, just specify `metrics: true`. See more on the [configuration reference](/reference/configuration.md#metrics) page.

Reviewer: ```suggestion

All the configurations are optional, to use the default, just specify `metrics: true`. See more on the [configuration reference](/reference/configuration.md#metrics) page.
```

> ** Please use environment variables in configuration to avoid exposing the credentials in the configuration file.** See more on the [configuration reference](/reference/configuration.md#environment-variable-placeholders).

Reviewer: ```suggestion
:::caution

Use [environment variable placeholders](/reference/configuration.md#environment-variable-placeholders) in your Platformatic DB configuration file to avoid exposing credentials.

:::
```

([Docusaurus admonitions](https://docusaurus.io/docs/markdown-features/admonitions))

## Grafana Configuration

Let's see how we can configure Grafana to chart some Platformatics metrics from Prometheus. 
Change the `docker-compose.yml` to add a `grafana` service:

```yaml
version: "3.7"
services:

  prometheus:
    image: prom/prometheus:latest
    volumes:
      - prometheus_data:/prometheus
      - ./prometheus/prometheus.yml:/etc/prometheus/prometheus.yml
    command:
      - '--config.file=/etc/prometheus/prometheus.yml'
    ports:
      - '9090:9090'

  grafana:
    image: grafana/grafana:latest
    volumes:
      - grafana_data:/var/lib/grafana
    environment:
      - GF_SECURITY_ADMIN_PASSWORD=pleasechangeme
    depends_on:
      - prometheus
    ports:
      - '3000:3000'

volumes:
  prometheus_data: {}
  grafana_data: {}

```

In Grafana, select `Configuration` -> `Data Sources` -> `Add Data Source`, and select Prometheus.
In the URL field, specify the URL of the Prometheus server, e.g. `http://prometheus:9090` (the name of the service in the `docker-compose` file), then Save & Test.

> Now we can create a dashboard, and add a panels to it. Select the Prometheus data source, and add queries. You should see the metrics exposed by Platformatic.

Reviewer: ```suggestion
Now we can create a dashboard and add panels to it. Select the Prometheus data source, and add queries. You should see the metrics exposed by Platformatic.
```
